### PR TITLE
JENKINS-68137 - Enable custom icon symbols for plugins

### PR DESCRIPTION
**Pull request by janfaracik**
See [JENKINS-68137](https://issues.jenkins.io/browse/JENKINS-68137) (but not really as it's just a lesser duplicate of this and only used for linking related issues)

This PR intends to add support for custom symbols so that plugin developers can add their own symbols for use in their plugins. I have an initial rough implementation (this PR) but I think there's still points to discuss.

1. Is this is the best way to approach loading the custom symbols/classloading?
2. ~What happens if a plugin has the same named symbol as a Jenkins Symbol?~ - namespaced
3. ~What happens if Plugin A has the same named symbol as Plugin B?~ - namespaced
4. Do Custom Symbols work on all pages?

Consumer: https://github.com/jenkinsci/ui-samples-plugin/pull/14

<details>
<summary>Docs</summary>

image

</details>

Link which will work for a bit:
https://ci.jenkins.io/job/Plugins/job/ui-samples-plugin/job/PR-14/35/artifact/io/jenkins/plugins/design-library/2.1-rc116.859da_e2f3952/design-library-2.1-rc116.859da_e2f3952.hpi

### Proposed changelog entries

* Developers can now use custom symbols in their plugins (link to Symbols docs)

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@timja @uhafner 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by timja** [comment]
can you squash the commits or mind if I do?

**Comment by timja** [comment]
> Is this is the best way to approach loading the custom symbols/classloading?

it looks okay without digging into the other issues

> What happens if a plugin has the same named symbol as a Jenkins Symbol?

probably not something we want to support although maybe some theme plugin may want to overwrite it.

> What happens if Plugin A has the same named symbol as Plugin B?

should be tested, possibly symbols should be registered rather than this magical loading.

> Do Custom Symbols work on all pages?

I suspect so, credentials-plugin context menu is where I had the most difficulty it has a lot of the Core API copied so may be hard to adjust, can likely have additional follow-ups if this isn't enough

**Review by uhafner — COMMENTED** [review]
The code looks pretty simple. What I don't understand. How does core know if I am using an image (current approach) or a symbol? How do I register new symbols?

**Comment by janfaracik** [comment]
> can you squash the commits or mind if I do?

For the life of me I can't figure out a way to do it cleanly, would be amazing if you could (and if you could tell me how ya did it 😃) - Cheers!

**Comment by janfaracik** [comment]
> The code looks pretty simple. What I don't understand. How does core know if I am using an image (current approach) or a symbol? How do I register new symbols?

It'll work much the same way as the current Jenkins Symbols, you have to use `<l:icon src="symbol-*" />` with `*` being the name of the symbol you want to use, eg `symbol-cloud`. As long as your icon is in `{plugin-root}/src/main/resources/images/symbols` it should just work. 

I'll update the design library plugin to have all this info with examples 👍

File: core/src/main/resources/lib/hudson/ballColorTd.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/hudson/ballColorTd.jelly:61]: We've generally used `clazz` in Jelly for this (with one painful exception IIRC).

**Comment by daniel-beck** [comment]
> What happens if Plugin A has the same named symbol as Plugin B?

How about a method to prevent overlap, e.g. automatically requiring that the plugin ID is the prefix of the symbol name?

**Comment by timja** [comment]
> > can you squash the commits or mind if I do?
> 
> For the life of me I can't figure out a way to do it cleanly, would be amazing if you could (and if you could tell me how ya did it 😃) - Cheers!

I used option 1 from this:
https://blog.oddbit.com/post/2019-06-17-avoid-rebase-hell-squashing-wi/

Previously (without knowing about the above technique) I've generated a patch file from a diff between branch and master.
Resetted to master and applied the patch creating a new commit.

File: core/src/main/resources/lib/hudson/ballColorTd.jelly
Comment (timja) [reply core/src/main/resources/lib/hudson/ballColorTd.jelly:61]: this isn't a new feature it's the icon class specification that's always been there

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:139]: we may want to cache this, not sure, any thoughts @daniel-beck?

**Comment by timja** [comment]
> > What happens if Plugin A has the same named symbol as Plugin B?
> 
> How about a method to prevent overlap, e.g. automatically requiring that the plugin ID is the prefix of the symbol name?

done, haven't adapted ui-samples yet but I checked in the debugger that it's looking for the right name

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:109]: I'm wondering if we should add another arg to the spec which is plugin-$pluginname and then use uberclassloader to load it if pluginname is set?

Comment (uhafner) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:109]: Would it also make sense to provide an API plugin for ionic icons in the same way I created an API plugin for Font Awesome icons? Then plugin authors would get automatically a simple way to use new, modern and consistent icons. Otherwise we might get a new mix of different icon styles again. Additionally, searching for consistent icons that use the correct SVG attributes might be cumbersome as well for plugin authors.

Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:109]: I think that would need to core as they are used here?

not a separate API plugin?

I've got something mostly working for the above, will prob finish it up tomorrow morning and push

Comment (uhafner) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:109]: > I think that would need to core as they are used here?
> not a separate API plugin?
 
Yes, that would be even better. But I would expect that such a library is too big for core so I did not even ask for it 😁 

> I've got something mostly working for the above, will prob finish it up tomorrow morning and push

😍

Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:109]: by above I meant namespacing plugins properly not the api plugin but will see if I have time for htat

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [inline core/src/main/java/org/jenkins/ui/icon/IconSet.java:102]: this isn't working. 😢 

I just get null when trying to use the plugins classloader that the plugin manager holds.

@janfaracik 's original design of passing the calling class won't work as we would want to have things like credentials plugin supplying icons that other plugins use, and library plugins like font-awesome, so we need to use the appropriate plugin classloader.

@jglick or @daniel-beck any idea what I'm doing wrong here?
Example plugin using this change: https://github.com/janfaracik/ui-samples-plugin/tree/overhaul

(core icons load fine as that just falls back to IconSet classloader)

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java]: I thought I had a nice clean plan, but the classloading isn't working 😢 

https://github.com/jenkinsci/jenkins/pull/6395/files#r835055749

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (jglick) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java:102]: Are you running inside a test or `hpi:run`?

**Comment by jglick** [comment]
(adjusting title so as not to confuse this with `@Symbol`)

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java]: `hpi:run`, this seems to fix it [`3137a98` (#6395)](https://github.com/jenkinsci/jenkins/pull/6395/commits/3137a98b7124b397d5aad1f26aa3ee313bdc6e72)

File: core/src/main/java/org/jenkins/ui/icon/IconSet.java
Comment (timja) [reply core/src/main/java/org/jenkins/ui/icon/IconSet.java]: resolved now

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2324]: TODO add test coverage

**Comment by timja** [comment]
I think the current implementation covers all the issues raised, any thoughts?

**Comment by timja** [comment]
Do we want to continue to maintain https://www.jenkins.io/doc/developer/views/symbols/ or should we point developers to the design library?

**Comment by NotMyFault** [comment]
> Do we want to continue to maintain https://www.jenkins.io/doc/developer/views/symbols/ or should we point developers to the design library?

Preferably, we cover it in the design lib instead of maintaining it in two occasions. The library is what we want to promote in the long run.

**Comment by timja** [comment]
This lgtm pending ATH and bom results

**Comment by timja** [comment]
I've added some initial docs:

Review / suggestions appreciated in https://github.com/jenkinsci/ui-samples-plugin/pull/14

image

**Review by timja — APPROVED** [review]
ATH + Bom  passed

Docs added

Looks good in the reference implementation

LGTM

----

Once this is released I think we can merge the design library PR, release that and then change the jenkins.io docs to point at it.
I'll create a draft PR for it later on

**Comment by timja** [comment]
> Do Custom Symbols work on all pages

Translations of symbols don't work on the Context Menu, and there doesn't seem to be a nice API to use them there.
If you can get the icon's XML you can add it in a 5 arg constructor or so but that's not very use friendly

e.g. see this in the Design Library:
```
.add(new MenuItem().withContextRelativeUrl("/").withIconClass("icon-up").withDisplayName("top page"));
```

Not really related to this PR though and should just have a targeted follow up for it

File: core/src/test/java/hudson/FunctionsTest.java
Comment (uhafner) [inline core/src/test/java/hudson/FunctionsTest.java:383]:
```suggestion
        String result = Functions.extractPluginNameFromIconSrc("");
```

File: core/src/test/java/hudson/FunctionsTest.java
Comment (uhafner) [inline core/src/test/java/hudson/FunctionsTest.java:389]: I must admit that from the code and test I do not actually understand how the input parameter should look like? Are these always two space separated String values? The first one the icon, the second one the plugin? Maybe it would be helpful if this is documented somewhere. Or is this explained in the design library?

**Review by uhafner — APPROVED** [review]

File: core/src/test/java/hudson/FunctionsTest.java
Comment (timja) [reply core/src/test/java/hudson/FunctionsTest.java:389]: It is in the design library

**Review by NotMyFault — APPROVED** [review]
🎉

**Comment by basil** [comment]
I think there should be a Jira ticket associated with this PR. If and when any regressions caused by this PR are reported in Jira, they could then be linked in Jira as "caused by" the Jira ticket associated with this PR.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by basil — APPROVED** [review]

**Comment by basil** [comment]
Causes [JENKINS-68270](https://issues.jenkins.io/browse/JENKINS-68270).

File: core/src/main/resources/lib/hudson/ballColorTd.jelly
Comment (basil) [reply core/src/main/resources/lib/hudson/ballColorTd.jelly:61]: @janfaracik This line (and this line alone) seems to cause [JENKINS-68270](https://issues.jenkins.io/browse/JENKINS-68270). Reverting this line to its old value fixes JENKINS-68270 for me.

Comment (basil) [reply core/src/main/resources/lib/hudson/ballColorTd.jelly:61]: I can confirm that the old icon class spec of `gitlab-logo icon-lg` works, while the new icon class spec of `gitlab-logo` (as of the above change) cannot be found in `iconsByCSSSelector`.

**Comment by basil** [comment]
Also causes [JENKINS-68296](https://issues.jenkins.io/browse/JENKINS-68296).

**Pull request merged**
